torch/lib/THD/base/init_methods/InitMethodTCP.cpp

```diff
@@ -202,9 +202,29 @@ std::vector<MulticastMessage> getMessages(struct sockaddr* addr, rank_type world
   return unpacked_msgs;
 }
 
-InitMethod::Config initTCPMaster(struct sockaddr* addr) {
-  // TODO
-  throw std::runtime_error("non-multicast tcp initialization not supported");
+
+InitMethod::Config initTCPMaster(std::string address, std::string str_port,
+                                 rank_type world_size, int assigned_rank) {
+  InitMethod::Config config;
+  if (assigned_rank == -1) {
+    throw std::invalid_argument("tcp:// method with non-multicast addresses "
+                                "requires manual rank assignment");
+  }
+
+  config.rank = convertToRank(assigned_rank);
+  auto port = convertToPort(std::stoul(str_port));
+  if (config.rank == 0) {
+    config.master.world_size = world_size;
+    config.master.listen_port = port;
+    std::tie(config.master.listen_socket, std::ignore) = listen(config.master.listen_port);
+    config.public_address = discoverWorkers(config.master.listen_socket, config.master.world_size);
```

Comment:
it's better to refer to `world_size` and `port` without taking them from `config.master`. It makes code shorter and more readable